osquery/tables/events/windows/powershell_events.cpp

```diff
@@ -6,212 +6,355 @@
-  if (eid != kScriptBlockLoggingEid) {
-    return Status::success();
+Status PowershellEventSubscriber::parseScriptMessageEvent(
+    boost::optional<Context::ScriptMessage>& script_message_opt,
+    const boost::property_tree::ptree& event) {
+  script_message_opt = {};
+
+  // Get the event timestamp from the Event.System object
+  Context::ScriptMessage output;
+  output.osquery_time = std::time(nullptr);
+
+  auto time_created_obj_opt =
+      event.get_child_optional("Event.System.TimeCreated");
+
+  if (!time_created_obj_opt) {
+    return Status::failure(
+        "The Event.System.TimeCreated path was not accessible in the XML event "
+        "object");
+  }
+
+  const auto& time_created_obj = time_created_obj_opt.value();
+
+  output.event_time = time_created_obj.get("<xmlattr>.SystemTime", "");
+  if (output.event_time.empty()) {
+    return Status::failure(
+        "The SystemTime attribute in the TimeCreated object is missing");
+  }
+
+  // Parse the rest of the object
+  auto event_data_obj_opt = event.get_child_optional("Event.EventData");
+  if (!event_data_obj_opt) {
+    return Status::failure(
+        "The Event.EventData path was not accessible in the XML event object");
   }
 
-  Row results;
-  for (const auto& node : ec->eventRecord.get_child("Event", pt::ptree())) {
-    if (node.first == "System" || node.first == "<xmlattr>") {
+  const auto& event_data_obj = event_data_obj_opt.value();
+
+  std::size_t field_count{0U};
+  bool malformed_field{false};
+
+  for (const auto& p : event_data_obj) {
+    const auto& node = p.second;
+
+    auto field_name = node.get("<xmlattr>.Name", "");
+    if (field_name.empty()) {
+      malformed_field = true;
       continue;
     }
-    // #4357: This should make use of RapidJSON
-    parseTree(node.second, results);
+
+    auto field_string_value = node.data();
+    auto field_integer_value_exp = tryTo<std::size_t>(field_string_value);
+
+    if (field_name == "MessageNumber") {
+      if (field_integer_value_exp.isError()) {
+        ++malformed_field;
+        continue;
+      }
+
+      output.message_number = field_integer_value_exp.take();
+      ++field_count;
+
+    } else if (field_name == "MessageTotal") {
+      if (field_integer_value_exp.isError()) {
+        ++malformed_field;
+        continue;
+      }
+
+      output.expected_message_count = field_integer_value_exp.take();
+      ++field_count;
+
+    } else if (field_name == "ScriptBlockText") {
+      output.message = field_string_value;
+      ++field_count;
+
+    } else if (field_name == "ScriptBlockId") {
+      output.script_block_id = field_string_value;
+      ++field_count;
+
+    } else if (field_name == "Path") {
+      output.script_path = field_string_value;
+      ++field_count;
+
+      if (!output.script_path.empty()) {
+        output.script_name =
+            boost::filesystem::path(output.script_path).leaf().string();
+      }
+    }
```

Comment:
Hmm. I can't _quite_ see it, but really feels like this whole block could be abstracted into some sort of dictionary-jump table with `field_name` as the key, or make use of lambdas, as we seem to have a fizz-buzz-bang happening. Just wondering if there might be a clever way to optimize this, but I'm not super clever :)